DigitalLearningSolutions.Web/Controllers/TrackingSystem/Delegates/GroupDelegatesController.cs

```diff
@@ -0,0 +1,217 @@
+        [HttpPost("Add/{delegateId:int}")]
+        public IActionResult AddDelegate(int groupId, int delegateId)
+        {
+            var delegateUser = userService.GetDelegateUserById(delegateId);
+            if (delegateUser == null)
+            {
+                return NotFound();
+            }
+
+            var adminId = User.GetAdminId();
+
+            var newDetails = new MyAccountDetailsData(
+                null,
+                delegateUser.Id,
+                string.Empty,
+                delegateUser.FirstName!,
+                delegateUser.LastName!,
+                delegateUser.EmailAddress!,
+                null
+            );
+
+            groupsService.AddDelegateToGroup(
+                delegateUser!.Id,
+                groupId,
+                0
+            );
+
+            groupsService.EnrolDelegateOnGroupCourses(
+                delegateUser!,
+                newDetails,
+                groupId,
+                adminId
+            );
+
+            return RedirectToAction("ConfirmDelegateAdded", new { groupId, delegateId });
+        }
+
+        [HttpGet("Add/{delegateId:int}/Confirmation")]
+        public IActionResult ConfirmDelegateAdded(int groupId, int delegateId)
+        {
+            var delegateUser = userService.GetDelegateUserById(delegateId);
+            if (delegateUser == null)
+            {
+                return NotFound();
+            }
+
+            var centreId = User.GetCentreId();
+            var groupName = groupsService.GetGroupName(groupId, centreId);
+
+            var model = new ConfirmDelegateAddedViewModel(delegateUser.FullName, groupName!, groupId);
+            return View(model);
+        }
+
+        [HttpGet("Add/SelectDelegate/AllItems")]
+        public IActionResult SelectDelegateAllItems(int groupId)
+        {
+            var centreId = User.GetCentreId();
+            var jobGroups = jobGroupsService.GetJobGroupsAlphabetical();
+            var customPrompts = centreCustomPromptHelper.GetCustomPromptsForCentre(centreId);
+            var delegateUsers = userService.GetDelegatesNotRegisteredForGroupByGroupId(groupId, centreId);
+
+            var model = new SelectDelegateAllItemsViewModel(
+                delegateUsers,
+                jobGroups,
+                customPrompts,
+                groupId
+            );
+
+            return View(model);
+        }
+
+
+        [HttpGet]
+        [Route("{delegateId:int}/Remove")]
+        public IActionResult RemoveGroupDelegate(int groupId, int delegateId)
+        {
+            var centreId = User.GetCentreId();
+            var groupName = groupsService.GetGroupName(groupId, centreId);
+            var groupDelegates = groupsService.GetGroupDelegates(groupId).ToList();
+            var delegateUser = groupDelegates.SingleOrDefault(gd => gd.DelegateId == delegateId);
+
+            if (groupName == null || delegateUser == null)
+            {
+                return NotFound();
+            }

```

Comment:
I feel like we could have a filter that verifies a user is in a particular group when the request includes a group and delegate id, but that might be overkill.

Reply:
I know what you mean. Although I would prefer to wait and see if there became a greater need for it.